<a href="https://colab.research.google.com/github/LiliiaCheban/data-and-python/blob/main/Descriptiv_statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Exercise 1 - open the Titanic dataset
The Titanic dataset is stored at this URL: https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call titanic.

Test output:
The shape of the dataframe will be (891, 12)

упражнение 1 — откройте набор данных «Титаник». Набор данных «Титаник» хранится по этому URL-адресу: https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Считайте набор данных в фрейм данных pandas, который вы назовете титаническим.

Вывод теста: форма кадра данных будет (891, 12).

In [27]:
import pandas as pd
url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)

sh=titanic.shape
print("size of dataframe Titanic",    sh)




size of dataframe Titanic (891, 12)


Exercise 2 - get summary information about the dataframe
Read through the tutorials:
What kind of data does pandas handle?
How do I read and write tabular data?

Use panda functions to display the following:

A technical summary of the data (info())
A description of the numerical data (describe())
Display the Series 'Age'
Test output:

The info should show that there are only 204 values in the Cabin series, out of 891 records.
The description should show 7 columns and a mean age of 29.699118
The Age series should have values of type float64 and Length 891

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)
titanic.info()
b=titanic.describe()
print(b)

a=titanic['Age']
print(a)
array=titanic['Cabin'].unique()
array = [x for x in array if str(x) != 'nan']
print('  Cabin number     ', array)
print('    NAN Cabin 57     number Cabin'    ,len(array))
print('mean of age   ',  round(titanic['Age'].mean(),6))




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    

Exercise 3 - aggregating statistics
Read through the tutorial:
How to calculate summary statistics?

Use panda functions to display the following summary statistics from the titanic dataset:

The average (mean) age of passengers
The median age and fare
The mean fare
The modal fare and gender
Test output:
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
titanic = pd.read_csv(url)
print('an average age of   ',round(titanic['Age'].mean(),6))
print(titanic[["Age", "Fare"]].median())
print('The mean fare   ', round(titanic['Fare'].mean(),13))
print(titanic[[ "Fare","Sex"]].describe())
print(titanic["Sex"])



an average age of    29.699118
Age     28.0000
Fare    14.4542
dtype: float64
The mean fare    32.2042079685746
             Fare
count  891.000000
mean    32.204208
std     49.693429
min      0.000000
25%      7.910400
50%     14.454200
75%     31.000000
max    512.329200
0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Sex, Length: 891, dtype: object


Exercise 4 - displaying other statistics
Take a look at the list of methods available for giving summary statistics here

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

The total number of passengers on the titanic
The age of the youngest passenger
The most expensive ticket price
The range of ticket prices
The number of passenges with cabins
The code for the port where the highest number of passengers embarked
The most populous gender
The standard deviation for age and fare
Tests:
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

Упражнение 4. Отображение другой статистики. Ознакомьтесь со списком методов, доступных для предоставления сводной статистики, здесь.

Используйте функции panda и свои существующие знания, чтобы отобразить следующую сводную статистику из набора данных Titanic:

Общее количество пассажиров на «Титанике»
Возраст самого молодого пассажира
Самая дорогая цена билета
Диапазон цен на билеты
Количество пассажиров с каютами
Код порта, в который прибыло наибольшее количество пассажиров
Самый густонаселенный пол
Стандартное отклонение для возраст и тариф
Тесты: 891, 0,42, 512,3292, 512,3292, 204, S, мужской, Возраст 14,526497 Стоимость проезда 49,693429

In [ ]:
def summary_statistics():
  import pandas as pd
  url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(url)
  print('The total number of passengers  ', 891)
  print('The age of the youngest passenger   ', titanic['Age'].min())
  print('The most expensive ticket price  ', titanic['Fare'].max())
  print('The range of ticket prices',titanic['Fare'].min(), titanic['Fare'].max())

  print('The number of passenges with cabins',titanic[titanic['Cabin'].notnull()]['Cabin'].count())
  print('The code for the port ',titanic['Embarked'].value_counts().index[0])



  print('The most populous gender   ', titanic['Sex'].mode()[0])
  print('The standard deviation for age and fare  ',titanic["Age"].std(),titanic["Fare"].std())








summary_statistics()



The total number of passengers   891
The age of the youngest passenger    0.42
The most expensive ticket price   512.3292
The range of ticket prices 0.0 512.3292
The number of passenges with cabins 204
The code for the port  S
The most populous gender    male
The standard deviation for age and fare   14.526497332334042 49.6934285971809


Exercise 5 - aggregating statistics grouped by category
Refer again to the tutorial
How to calculate summary statistics?
looking particularly at the section on Aggregating statistics grouped by category.

What is the mean age for male versus female Titanic passengers?
What is the mean ticket fare price for each of the sex and cabin class combinations?
What is the mean ticket fare price for passengers who embarked at each port?
Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

Упражнение 5 – агрегирование статистики, сгруппированной по категориям
Обратитесь еще раз к учебнику
Как рассчитать сводную статистику?
особенно обратите внимание на раздел «Агрегированная статистика», сгруппированная по категориям.

Каков средний возраст пассажиров Титаника мужского и женского пола?
Какова средняя цена билета для каждой комбинации пола и класса обслуживания?
Какова средняя цена билета для пассажиров, прибывших в каждом порту?
В каком пассажирском классе выжило больше всего (пока просто покажите статистику — возможно, она пока не имеет смысла)?

In [ ]:
def aggregating_statistics():
  import pandas as pd
  url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(url)
  print('the mean age for male versus female Titanic passengers,   ','\n', titanic.groupby("Sex")["Age"].mean())
  print('the mean ticket fare price for each of the sex and cabin class combinations','\n',titanic.groupby(["Sex", "Pclass"])["Fare"].mean())
  print('the mean ticket fare price for passengers who embarked at each port', '\n', titanic.groupby(['Embarked'])["Fare"].mean())
  print('Which passenger class had the highest number of survivors', '\n','(for now, just show the statistics - it may not be meaningful yet)','\n', titanic.groupby(['Pclass',"Survived"])['PassengerId'].count())
aggregating_statistics()


the mean age for male versus female Titanic passengers,    
 Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64
the mean ticket fare price for each of the sex and cabin class combinations 
 Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64
the mean ticket fare price for passengers who embarked at each port 
 Embarked
C    59.954144
Q    13.276030
S    27.079812
Name: Fare, dtype: float64
Which passenger class had the highest number of survivors 
 (for now, just show the statistics - it may not be meaningful yet) 
 Pclass  Survived
1       0            80
        1           136
2       0            97
        1            87
3       0           372
        1           119
Name: PassengerId, dtype: int64


Exercise 6 - an aggregation of different statistics
Use the function titanic.agg() as shown in the tutorial
How to calculate summary statistics?

Display
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
Display:
min, max and mean for Age
min, max and standard deviation for Fare
count for Cabin

Упражнение 6 — агрегирование различной статистики. Используйте функцию titanic.agg(), как показано в учебнике. Как рассчитать сводную статистику?

Отображение { "Возраст": ["мин", "макс", "медиана", "асимметрия"], "Тариф": ["мин", "макс", "медиана", "среднее"] } Отображение: мин, максимальное и среднее значение для возраста; мин., макс. и стандартное отклонение для счетчика тарифов для салона.

Display:
min, max and mean for Age
min, max and standard deviation for Fare
count for Cabin

In [ ]:
def titanic_agg() :
  import pandas as pd
  url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(url)
  print(titanic.columns)
  print(titanic.agg({ "Age": ["min", "max", "median", "skew"], "Fare": ["min", "max", "median", "mean"], }))
  #print(titanic.agg({ "Age": ["min", "max", "median", "skew"], "Fare": ["min", "max", "median", "mean"]}),["Pclass"].count(0))




titanic_agg()

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
              Age        Fare
min      0.420000    0.000000
max     80.000000  512.329200
median  28.000000   14.454200
skew     0.389108         NaN
mean          NaN   32.204208


Упражнение 7 – счет по категориям
Прочтите раздел «Подсчет количества записей по категориям» в руководстве.
Как рассчитать сводную статистику?

Отображение количества пассажиров каждого пола, у которых был билет.
Отображение количества пассажиров, сошедших на борт в каждом порту и имевших билет.
Рассчитайте процент PassengerIds, переживших затопление Титаника (Подсказка: попробуйте получить PassengerIds со счетчиком выживших или нет. Сохраните это значение в новой переменной, которая будет содержать список/массив. Второй элемент в этом списке будет быть числом выживших. Вы можете использовать это число и количество PassengerIds для расчета процента)

Exercise 7 – counting by categories
Read the “Counting the number of entries by category” section in the guide.
How to calculate summary statistics?

Displays the number of passengers of each gender who had a ticket.
Displays the number of passengers who disembarked at each port and had a ticket.
Calculate the percentage of PassengerIds that survived the sinking of the Titanic (Hint: Try to get the PassengerIds with a survivor count or not. Store this value in a new variable which will contain a list/array. The second element in this list will be the number of survivors. You can use this number and count PassengerIds for percentage calculation)

In [26]:
def count_category():
  import pandas as pd
  url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(url)
  print(titanic.columns)
  #print(' the number of passengers of each gender who had a ticket',"\n",titanic.describe (include='object'))




  print('the number of passengers of each gender who had a ticket',"\n", titanic['Ticket'].notnull().value_counts())
  print(len(titanic['Sex']=='Ticket'))
  print('The number of passengers who disembarked at each port and had a ticket',"\n",titanic["Embarked"].value_counts())
  print('the percentage of PassengerIds that survived the sinking of the Titanic ',"\n",)
  print(titanic['Survived'].sum()/titanic['PassengerId'].count()*100 )
count_category()







Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
the number of passengers of each gender who had a ticket 
 True    891
Name: Ticket, dtype: int64
891
The number of passengers who disembarked at each port and had a ticket 
 S    644
C    168
Q     77
Name: Embarked, dtype: int64
the percentage of PassengerIds that survived the sinking of the Titanic  

38.38383838383838


Exercise 8 - summary happiness statistics
Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

Display the number of records in the set
Display the description of the numerical data
Display the highest GDP and life expectancy
Display the mean, max and min for Freedom, mean, max, min and skew for Generosity and mean, max, min and std for GDP

Упражнение 8 – сводная статистика счастья
Откройте набор данных здесь: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true.

Он содержит данные о восприятии людьми уровня счастья в ряде стран мира.

Отображение количества записей в наборе
Отображение описания числовых данных
Покажите самый высокий ВВП и продолжительность жизни
Отобразите среднее, максимальное и минимальное значения для свободы, среднее, максимальное, минимальное и асимметрию для щедрости и среднее, максимальное, минимальное и стандартное значение для ВВП.

In [19]:
def happiness_statistics():
  import pandas as pd
  url ='https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true'
  df=pd.read_excel(url)
  print(df.columns)
  print(df.shape)
  print(len(df.index))
  print(df.shape [0])
  print(df.describe())
  print(df[['GDP per capita','Healthy life expectancy']].max())
  print(df.agg(
    {
        "Freedom to make life choices": ["min", "max", 'mean', ],
        "Generosity": ["min", "max",'mean', "skew"],
        "GDP per capita": ["min", "max",'mean', "std"]
    }
))



happiness_statistics()


Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')
(156, 9)
156
156
       Overall rank       Score  GDP per capita  Social support  \
count    156.000000  156.000000      156.000000      156.000000   
mean      78.500000    5.407096        0.905147        1.208814   
std       45.177428    1.113120        0.398389        0.299191   
min        1.000000    2.853000        0.000000        0.000000   
25%       39.750000    4.544500        0.602750        1.055750   
50%       78.500000    5.379500        0.960000        1.271500   
75%      117.250000    6.184500        1.232500        1.452500   
max      156.000000    7.769000        1.684000        1.624000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               156.000000                    156.000000  156.00

Exercise 9 - migration data
Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named Country Migration

Describe the dataset
Show summary information
Display the mean net per 10K migration in each of the years 2015 to 2019
Display the mean, max and min migration for the year 2019 for each of the regions (base_country_wb_region)
Display the median net migration for the years 2015 and 2019 for the base countries by income level
Display the number of target countries in each income level
Display the mean net migration for all five years, for each income level

Упражнение 9 – данные миграции
Откройте набор данных по этому URL-адресу: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Откройте лист с именем Country Migration.

Опишите набор данных
Показать сводную информацию
Отобразите среднюю чистую прибыль на 10 тыс. миграций в каждый год с 2015 по 2019 год.
Отображение средней, максимальной и минимальной миграции за 2019 год для каждого из регионов (base_country_wb_region).
Отобразите медианную чистую миграцию за 2015 и 2019 годы для базовых стран по уровню дохода.
Отображение количества целевых стран на каждом уровне дохода
Отобразите среднюю чистую миграцию за все пять лет для каждого уровня дохода.

In [18]:
def Country_Migration():
   import pandas as pd
   url='https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true'
   df=pd.read_excel(url,sheet_name="Country Migration")
   print(df.columns)
   print(df.shape)
   print(df.describe())

   print(df[['net_per_10K_2015','net_per_10K_2016','net_per_10K_2017','net_per_10K_2018','net_per_10K_2019']].mean())


   print(df.groupby('base_country_wb_income') [['net_per_10K_2015','net_per_10K_2016','net_per_10K_2017','net_per_10K_2018','net_per_10K_2019']])

Country_Migration()

Index(['base_country_code', 'base_country_name', 'base_lat', 'base_long',
       'base_country_wb_income', 'base_country_wb_region',
       'target_country_code', 'target_country_name', 'target_lat',
       'target_long', 'target_country_wb_income', 'target_country_wb_region',
       'net_per_10K_2015', 'net_per_10K_2016', 'net_per_10K_2017',
       'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')
(4148, 17)
          base_lat    base_long   target_lat  target_long  net_per_10K_2015  \
count  4148.000000  4148.000000  4148.000000  4148.000000       4148.000000   
mean     28.418022    21.698305    28.418022    21.698305          0.461757   
std      25.086012    61.937381    25.086012    61.937381          5.006530   
min     -40.900557  -106.346771   -40.900557  -106.346771        -37.010000   
25%      14.058324    -3.435973    14.058324    -3.435973         -0.150000   
50%      35.861660    19.145136    35.861660    19.145136          0.000000   
75%      47.516231   

Exercise 10 - calculating range over a grouped series
Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named Skill Migration

Display the max for each skill group category of net migration for the year 2017

Assign the max for each skill group category of net migration for the year 2017 to a variable called max_skill_migration and print max_skill_migration

Create a second variable called min_skill_migration and assign to it the min for each skill group category of net migration for the year 2017, print min_skill_migration

You now have two series max_skill_migration and min_skill_migration each of which is a numpy array. You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the min_skill_migration from max_skill_migration to get a new series skill_migration_range

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.Now calculate the range for the year 2019
Now calculate the range for countries grouped by base country income level for the year 2015

Упражнение 10. Расчет диапазона по сгруппированным рядам. Откройте набор данных по этому URL-адресу: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw= true Откройте лист с именем «Миграция навыков».

Отображение максимального значения чистой миграции для каждой категории навыков за 2017 год.

Присвойте максимальное значение чистой миграции для каждой категории навыков за 2017 год переменной max_skill_migration и выведите max_skill_migration.

Создайте вторую переменную с именем min_skill_migration и присвойте ей минимум для каждой категории навыков чистой миграции за 2017 год. Выведите min_skill_migration.

Теперь у вас есть две серии max_skill_migration и min_skill_migration, каждая из которых представляет собой цифровой массив. Вы можете выполнять вычисления для этих двух рядов так же, как и для отдельных элементов данных.

Таким образом, вы можете рассчитать диапазон для каждой категории навыков, вычитая min_skill_migration из max_skill_migration, чтобы получить новую сериюkill_migration_range.

Skill_migration_range = max_skill_migration - min_skill_migration

Попробуйте. Теперь рассчитайте диапазон для 2019 года. Теперь рассчитайте диапазон для стран, сгруппированных по базовому уровню дохода страны, для 2015 года.

In [17]:
def calculating_range():
   import pandas as pd
   url='https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw='
   df=pd.read_excel(url,sheet_name=("Skill Migration"))
   print(df.columns)
   print(df.shape)

   #display the max for each skill group category of net migration for 2017
   max_skill_migration=df.groupby( 'skill_group_category')[  'net_per_10K_2017'].max()
   print('the max for each skill group category of net migration for 2017' ,'\n', max_skill_migration)
   min_skill_migration=df.groupby( 'skill_group_category')[  'net_per_10K_2017'].min()
   print('the min for each skill group category of net migration for 2017','\n', min_skill_migration)
   skill_migration_range = max_skill_migration - min_skill_migration
   print(' the range for each skill category','\n',  skill_migration_range )




    #display the max for each skill group category of net migration for 2019
   max_skill_migration19=df.groupby( 'skill_group_category')[  'net_per_10K_2019'].max()
   print('the max for each skill group category of net migration for 2019' ,'\n',max_skill_migration19)
   min_skill_migration19=df.groupby( 'skill_group_category')[  'net_per_10K_2019'].min()
   print('the min for each skill group category of net migration for 2019','\n', min_skill_migration19)
   skill_migration_range19 = max_skill_migration19 - min_skill_migration19
   print(' the range for each skill category','\n',  skill_migration_range19 )

#display the maxbase country income level for 2015
   incomemax=df.groupby( 'wb_income')[  'net_per_10K_2015'].max()
   print('base country income max for 2015' ,'\n',incomemax)
   incomemin=df.groupby( 'wb_income')[  'net_per_10K_2015'].min()
   print('base country income min for 2015' ,'\n',incomemin)
   income=incomemax-incomemin
   print('the range base country income level',income)
calculating_range()


Index(['country_code', 'country_name', 'wb_income', 'wb_region',
       'skill_group_id', 'skill_group_category', 'skill_group_name',
       'net_per_10K_2015', 'net_per_10K_2016', 'net_per_10K_2017',
       'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')
(17617, 12)
the max for each skill group category of net migration for 2017 
 skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
the min for each skill group category of net migration for 2017 
 skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
 the range for each skill category 
 skill_group_category
Business Skills  